Storefront API delegate access token storage

rennyG opened the pull request:
<!-- Thank you for contributing! -->

### Description

Partially closes https://github.com/Shopify/shopify-dev/issues/23838

- explains using private variables for server tokens
- Updates deployment.md to highlight server tokens as something that Partners need to complete

### Additional context

Have I got this right?

Is the location for the content acceptable?

Will using `public access`, `public variable`, and `public token` be confusing for partners? 

## 🎩 / 🌀 
https://shopify-dev.hydrogen-s2s.ren-chaturvedi.us.spin.dev/custom-storefronts/hydrogen

---

### Before submitting the PR, please make sure you do the following:

- [x] Read the [Contributing Guidelines](https://github.com/shopify/hydrogen/blob/main/.github/contributing.md)
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`)
- [x] Update docs in this repository according to your change
- [ ] Run `yarn changeset add` if this PR cause a version bump based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)

blittle commented on docs/deployment.md:
```diff
@@ -18,10 +18,16 @@ Oxygen is Shopify's recommended deployment platform for Hydrogen storefronts. To
 To learn how to deploy your Hydrogen storefront to Netlify, refer to the [Hydrogen on Netlify](https://docs.netlify.com/integrations/frameworks/hydrogen/) documentation.
 
+> Caution:
+> To avoid rate limiting on server requests to the Storefront API, you need to create a delegate access token for server requests to the Storefront API and store it in the private environment variable `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`. [Learn how](#avoid-rate-limiting-in-production).
```
Ongoing discussion on removing `SHOPIFY` portion of the variable name

blittle commented on docs/deployment.md:
```diff
@@ -335,8 +347,43 @@ You can deploy your Hydrogen storefront to Cloudflare Workers, a serverless appl
+
+1. Create a [delegate access token](https://shopify.dev/apps/auth/oauth/delegate-access-tokens) for the Storefront API.
+
+1. [Store the token](https://vitejs.dev/guide/env-and-mode.html#env-files) in a private environment variable in `Oxygen.env`.
```
I don't think it makes sense to tell them to put it into `Oxygen.env`. If they aren't deploying to Oxygen, might as well put it wherever they want. I'd say just

> [Store the token](https://vitejs.dev/guide/env-and-mode.html#env-files) in a private environment variable: `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`

rennyG replied:
Look at making corollary changes to environment-variables.md

blittle commented on docs/deployment.md:
```diff
@@ -335,8 +347,43 @@ You can deploy your Hydrogen storefront to Cloudflare Workers, a serverless appl
+
+1. [Store the token](https://vitejs.dev/guide/env-and-mode.html#env-files) in a private environment variable in `Oxygen.env`.
+
+1. Set the private token using the global `Oxygen.env` object in the Hydrogen configuration:
```
Then update to mention just the variable, not the global `Oxygen.env`

rennyG replied:
I'm guessing corollary changes are required for the public variable too, correct? [For example](https://github.com/Shopify/hydrogen/pull/1994/files#diff-7f6254f34e124b6ff67f63aacb73b5252aa8e2b8afe712f48cdd3b32a24f908cR385)

blittle commented on docs/deployment.md:
````diff
@@ -335,8 +347,43 @@ You can deploy your Hydrogen storefront to Cloudflare Workers, a serverless appl
+    ```tsx
+    export default defineConfig({
+      privateStorefrontToken:
+        Oxygen?.env?.PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN,
````
And this would be updated too

rennyG replied:
Any delimiters or path variables required?

e.g. `{path}.{path}.PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`?

My vote would be to remain opinionated in the docs. This makes the docs simpler to write and maintain. We can always say that they can store it anywhere and that we're providing this as an example. What do you think? @blittle @mcvinci

mcvinci replied:
@rennyG I'm on board with referencing that they can be stored anywhere, and explicitly stating that we're providing an example.

mcvinci commented on docs/framework/analytics.md:
```diff
@@ -3,6 +3,8 @@ gid: 044c475e-f28g-444b-a984-26e7ebb8bec4
 title: Analytics
 description: Learn about the analytics support built into Hydrogen.
 ---
+> Caution:
+> Include the `storefrontId` prop in the Hydrogen config to avoid breaking the analytics dashboard in the Shopify admin. Learn more about [Hydrogen configuration](https://shopify.dev/custom-storefronts/hydrogen/framework/hydrogen-config).
```
Placing this prop-specific information this high up in the topic feels a bit out of context. Can we add this information as part of a step in the topic? 

We reference the `hydrogen.config.js` file as part of the second step here: https://shopify.dev/custom-storefronts/hydrogen/framework/analytics#send-analytics-data-from-the-server-side

rennyG replied:
What do you think about setting as a requirement, per deployment?

mcvinci replied:
Sounds like a good approach!

rennyG replied:
Ended up not going that route, because I thought the level at which the requirement sits might get missed, since the step to add the storefront ID is in one procedure, lower down on the page. I think I got the point across. I'm using comments to keep this bit of information close to the code. If/when you have time, could you take a look?

mcvinci commented on docs/deployment.md:
```diff
@@ -18,10 +18,16 @@ Oxygen is Shopify's recommended deployment platform for Hydrogen storefronts. To
 To learn how to deploy your Hydrogen storefront to Netlify, refer to the [Hydrogen on Netlify](https://docs.netlify.com/integrations/frameworks/hydrogen/) documentation.
 
+> Caution:
+> To avoid rate limiting on server requests to the Storefront API, you need to create a delegate access token and store it in the private environment variable `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`. [Learn more](#avoid-rate-limiting-in-production).
```
```suggestion
> To [avoid rate limiting](#avoid-rate-limiting-in-production) on server requests to the Storefront API, you need to create a delegate access token and store it in the private environment variable `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`.
```

mcvinci replied:
Instead of adding this note to each applicable deployment section, maybe we can simplify and add this in a single instance to the `Requirements` section in the topic?

rennyG replied:
Yep that works! I was assuming that people would pay more attention to it if it was colocated with the runtime section, but putting in requirements is tidier, easier to maintain, and is pretty clear imo

mcvinci commented on docs/deployment.md:
```diff
@@ -18,10 +18,16 @@ Oxygen is Shopify's recommended deployment platform for Hydrogen storefronts. To
 To learn how to deploy your Hydrogen storefront to Vercel Edge Functions, refer to the [Vercel Hydrogen template](https://vercel.com/new/clone?s=https%3A%2F%2Fgithub.com%2Fvercel%2Fvercel%2Ftree%2Fmain%2Fexamples%2Fhydrogen&template=hydrogen) documentation.
 
+> Caution:
+> To avoid rate limiting on server requests to the Storefront API, you need to create a delegate access token and store it in the private environment variable `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`. [Learn more](#avoid-rate-limiting-in-production).
```
```suggestion
> To [avoid rate limiting](#avoid-rate-limiting-in-production) on server requests to the Storefront API, you need to create a delegate access token and store it in the private environment variable `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`.
```

mcvinci replied:
Potential to remove this note and move the information up to the `Requirements` section.

rennyG replied:
Done

mcvinci commented on docs/deployment.md:
```diff
@@ -172,6 +178,10 @@ module.exports = function (request, response) {
 
+### Configure a delegate access token
+
+To avoid rate limiting on server requests to the Storefront API, you need to create a delegate access token for server requests, stored in a private environment variable. [Learn how](#avoid-rate-limiting-in-production).
```
```suggestion
To [avoid rate limiting](#avoid-rate-limiting-in-production) on server requests to the Storefront API, you need to create a delegate access token and store it in the private environment variable `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`.
```

mcvinci replied:
Potential to remove this note and move the information up to the `Requirements` section.

rennyG replied:
Done

mcvinci commented on docs/deployment.md:
```diff
@@ -202,7 +212,9 @@ You can deploy your project to any platform that supports Docker-based hosting,
     {% endcodeblock %}
 
-3. Run Docker inside your app directory by executing the following commands:
+1. Create and store a delegate access token for server requests to the Storefront API. [Learn how](#avoid-rate-limiting-in-production).
```
```suggestion
1. [Create and store a delegate access token](#avoid-rate-limiting-in-production) for server requests to the Storefront API.
```

mcvinci replied:
Potential to remove this step and move it up to the `Requirements` section.

mcvinci commented on docs/deployment.md:
````diff
@@ -335,8 +347,43 @@ You can deploy your Hydrogen storefront to Cloudflare Workers, a serverless appl
     ```
 
-6. Deploy your project with [Wrangler](https://developers.cloudflare.com/workers/cli-wrangler/install-update):
+1. Create and store a delegate access token for server requests to the Storefront API. [Learn how](#avoid-rate-limiting-in-production).
````
```suggestion
1. [Create and store a delegate access token](#avoid-rate-limiting-in-production) for server requests to the Storefront API.
```

mcvinci replied:
Potential to remove this step and move it up to the `Requirements` section.

mcvinci commented on docs/framework/environment-variables.md:
```diff
@@ -46,27 +55,30 @@ export default Component() {
+## Private variables
 
-### Private variables
+In Hydrogen, variables that are prefixed with `PRIVATE` in the `.env` file is treated as a server runtime variable in non-production environments.
```
```suggestion
In Hydrogen, variables that are prefixed with `PRIVATE` in the `.env` file are treated as a server runtime variable in non-production environments.
```

mcvinci commented on docs/deployment.md:
```diff
@@ -335,8 +347,43 @@ You can deploy your Hydrogen storefront to Cloudflare Workers, a serverless appl
+
+## Avoid rate limiting in production
+
+This is a necessary step if you are deploying to a non-Oxygen runtime. Learn [why this step is required](/https://shopify.dev/custom-storefronts/hydrogen/framework/environment-variables#store-server-tokens-as-private-variables).
```
```suggestion
This is a necessary step if you're deploying to a non-Oxygen runtime. Learn [why this step is required](/https://shopify.dev/custom-storefronts/hydrogen/framework/environment-variables#use-storefront-api-server-tokens).
```

mcvinci commented on docs/deployment.md:
```diff
@@ -335,8 +347,43 @@ You can deploy your Hydrogen storefront to Cloudflare Workers, a serverless appl
+
+1. [Store the token](https://vitejs.dev/guide/env-and-mode.html#env-files) in a private environment variable called `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`.
+
+1. In the Hydrogen configuration, set the private token using the variable `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`.
```
```suggestion
1. In the Hydrogen configuration file, set the private token using the variable `PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`.
```

rennyG replied:
Based on the Hydrogen docs repo, `hydrogen configuration` is valid. Happy to make this change here though, as personally I think it's clearer.

mcvinci commented on docs/deployment.md:
```diff
@@ -335,8 +347,43 @@ You can deploy your Hydrogen storefront to Cloudflare Workers, a serverless appl
+
+    {% endcodeblock %}
+
+1. In the Hydrogen configuration, set the `storefrontId` using the variable `PUBLIC_SHOPIFY_STOREFRONT_ID`.
```
```suggestion
1. In the Hydrogen configuration file, set the `storefrontId` using the variable `PUBLIC_SHOPIFY_STOREFRONT_ID`.
```

mcvinci reviewed (approved):
Looking good, @rennyG! I added few comments and suggestions throughout.

mcvinci commented on docs/deployment.md:
```diff
@@ -335,8 +337,51 @@ You can deploy your Hydrogen storefront to Cloudflare Workers, a serverless appl
+
+    {% endcodeblock %}
+
+    > Note:
```
Might be worthwhile updating this to a `Caution` to indicate the breakage possibility?

rennyG replied:
yep good call. I think that's how I had it originally. D'oh.

mcvinci commented on docs/framework/analytics.md:
```diff
@@ -324,6 +324,23 @@ To send analytics data from the server-side, complete the following steps:
+
+    {% endcodeblock %}
+
+    > Note:
```
```suggestion
    > Caution:
```

mcvinci commented on docs/framework/environment-variables.md:
```diff
@@ -46,27 +58,29 @@ export default Component() {
+## Private variables
 
-### Private variables
+In Hydrogen, variables that are prefixed with `PRIVATE` in the `.env` file are treated as a server runtime variable in non-production environments.
```
```suggestion
In Hydrogen, variables that are prefixed with `PRIVATE` in the `.env` file are treated as server runtime variables in non-production environments.
```

mcvinci commented on docs/framework/environment-variables.md:
```diff
@@ -94,7 +108,28 @@ app.use(hydrogenMiddleware({/* ... */}))
+
+Prevent rate-limiting on server requests to the Storefront API by storing server tokens as private variables.
+
+You need to authenticate server requests to the Storefront API with a [delegate access token](/apps/auth/oauth/delegate-access-tokens) that's stored in a private variable (`PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`) and referenced in the [Hydrogen configuration](https://shopify.dev/custom-storefronts/hydrogen/framework/hydrogen-config):
```
```suggestion
You need to authenticate server requests to the Storefront API with a [delegate access token](/apps/auth/oauth/delegate-access-tokens) that's stored in a private variable (`PRIVATE_SHOPIFY_STOREFRONT_API_TOKEN`) and reference the private variable in the [Hydrogen configuration](https://shopify.dev/custom-storefronts/hydrogen/framework/hydrogen-config):
```

mcvinci replied:
Not quite sure if I got this right -- feel free to tweak / adjust as needed!

rennyG replied:
Stored and referenced are parallel, so if this isn't a blocker I'll leave as is

rennyG reviewed (commented):
Thanks MV! I need to figure out a good method of content reuse for the Hydrogen repo. I know that partials will resolve when the changes are ported to shopify-dev, but I'd prefer to keep the Hydrogen repo content all in one place.

rennyG commented:
> @rennyG I just took an advanced look at these docs to try to start using the delegate token early. Thank you for getting these written!
> 
> I will say that I struggled to move through these steps with confidence, perhaps the delegate token documentation may be augmented to help folks know that they're doing the right thing. Here are some points where I got confused:
> 
> * I wasn't sure where to access the GraphQL API to create the delegate token. I assumed the GraphQL admin API and moved forward.
> * I had the Shopify Graphiql App installed in my storefront, so I used that and the documentation [here](https://shopify.dev/apps/auth/oauth/delegate-access-tokens) to generate a token
> * I removed the expiration but then dug into the `expires_in` documentation. It seems like the delegate token is... delegate to the token that created it? This is where I really lost confidence. What token did I use to create it? Was it the admin token associated with the GraphiQL app? Was it associated with me? What are the ways that this token will become invalid, thus causing a production bug?
> 
> I'd love some more details on how to migrate to this delegate token workflow!

This is great feedback, thank you! I'll take a closer look and aim to remove these pain points.

rennyG commented:
> > @rennyG I just took an advanced look at these docs to try to start using the delegate token early. Thank you for getting these written!
> > I will say that I struggled to move through these steps with confidence, perhaps the delegate token documentation may be augmented to help folks know that they're doing the right thing. Here are some points where I got confused:
> > 
> > * I wasn't sure where to access the GraphQL API to create the delegate token. I assumed the GraphQL admin API and moved forward.
> > * I had the Shopify Graphiql App installed in my storefront, so I used that and the documentation [here](https://shopify.dev/apps/auth/oauth/delegate-access-tokens) to generate a token
> > * I removed the expiration but then dug into the `expires_in` documentation. It seems like the delegate token is... delegate to the token that created it? This is where I really lost confidence. What token did I use to create it? Was it the admin token associated with the GraphiQL app? Was it associated with me? What are the ways that this token will become invalid, thus causing a production bug?
> > 
> > I'd love some more details on how to migrate to this delegate token workflow!
> 
> This is great feedback, thank you! I'll take a closer look and aim to remove these pain points.

I might have addressed some of the ambiguity yesterday in [this PR](https://github.com/Shopify/shopify-dev/pull/25924) against the dev docs repo. Shouldn't be an issue to provide some clarity on where the token is created and about its parent.  I can dig in on this Monday, but might lean on either @vixdug or @blittle to help me fill in the details.

rennyG commented:
Shipping these changes. I've got another PR in the works to clear up the last bullet point in @davidhousedev's feedback. Otherwise, I made some changes to the content in other issues, which hopefully cover the first and second points.
